### PR TITLE
ACLs in Swarm see #1199 (ReOpen for #1365)

### PR DESCRIPTION
**Pull request by doronp**
ACLs in Swarm see #1199 #1365 

Authorization (ACLs) support in Swarm as a first step towards multi tenancy. #1199

This PR includes (organized by Commits for convenience)

1 - Validation API & Default implementation- Approve/Disapprove/condition of 'Who wants to to what'?
2 - Handler API & Default implementation - Use labels to enforce decision of 1
3 - Feature flag --multiTenant to get this feature going (regular Swarm otherwise) 

(Re open PR for internal PM reasons)

### CODE REVIEW

**Comment by thaJeztah** [comment]
@vieux @aluzzardi should this be looked at together with https://github.com/docker/docker/issues/13697 and https://github.com/docker/docker/pull/17412 (@NathanMcCauley is currently working with them on that proposal)

**Comment by doronp** [comment]
@abronan @chanwit Done with Most of the comments (except Shell tests should be converted to BATS I think)
I agree that moving to godep should happen at a later phase.
Kindly put the tags from the old PR to this one and lets continue our work here.

**Comment by abronan** [comment]
@doronp done! :) But it seems that some of your commits are still not signed though :( Also would be better if you can squash some commits (like all those `Fix for golint` mostly). Especially if you plan to add some more on top.

@thaJeztah Thanks this is helpful! We'll take a look at it as this seems closely related.

**Comment by tiborvass** [comment]
FYI there's also docker/docker#17412 and docker/docker#17687 dealing with this.

**Comment by doronp** [comment]
@chanwit : Added BATS tests, however since the tests are run without the --multiTenant flag they fail and hence the build fails. Can you put the Flag on for those tests?
P.S. The tag milestone 1.1.0 did not move from the closed PR (#1365) if it matters.
P.P.S When squashing I older commit inside (not mine) - Should I override that or is that fine?
@tiborvass: Tnx.

**Comment by abronan** [comment]
@doronp Digging into this one. We are seeing the first bits being merged in docker with docker/docker#15365 (for `authZ` support with plugins). I wonder how we can think about an integration and use the support straight out of docker maybe? This would simplify the code tremendously on the Swarm side. WDYT?

**Comment by doronp** [comment]
@abronan. 

Indeed it would be neat to harness the docker side authZ support.
I am not completely sure what you had in mind but I can assume at least two additional approaches:

1 - Swarm will pass the requests to the engines in its cluster and those will communicate with the AuthZ back-end via the plugin - Swarm will then aggregate and handle the responses.

I suspect there will be lots of inefficacy in this approach - Swarm will have to access the engines directly every time - it would not be able to use the internal cluster data as to complete operation as it does today*. This will probably have tremendous performance impact especially on bigger cluster (which is a reasonable assumption if someone uses ACL at all)
for example:
	 ps -a - list all container will access all over the cluster - then filter out to show only the users  containers etc.

*We can keep the operations such as ps in Swarm and only make it do AuthZ  via the engines but it is still inefficient. 
		

2 - Take the AuthZ(+AuthN) plugin from docker and make swarm do that job itself - Swarm is just like an engine in that case

I am going on digging the possible implementation options.

let me know what you think.

**Comment by doronp** [comment]
@vieux @abronan @aluzzardi commit 497e328 is the plugin package import - It does not interfere with functionality and passes all tests. It will help a lot to review and merge it ASAP - better not mix it with the AuthZ plugin code (701d09e commit WIP)  Thanks.

**Comment by abronan** [comment]
ping @docker/swarm-maintainers

**Comment by doronp** [comment]
@vieux @dongluochen When can we start working on this (I think start with commit 497e328) ? 
I'd not want to upset Gordon;-)

**Comment by MHBauer** [comment]
Since what's added in the `refactor 402ec3ee` commit is moved in the later `add middleware f827edad` commit, I would rather it just went away.

Can't help with the godeps. This needs a rebase.

File: api/server.go [hunk 1]
```diff
@@ -34,9 +35,26 @@ func (d *dispatcher) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 // Server is a Docker API server.
 type Server struct {
-	hosts      []string
-	tlsConfig  *tls.Config
-	dispatcher *dispatcher
+	cfg          *Config
```
Comment (MHBauer) [inline api/server.go]: I see this is where the `Config` struct below on L#50 is used, but where is this pointer used?

File: api/server.go [hunk 2]
```diff
@@ -34,9 +35,26 @@ func (d *dispatcher) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+	hosts        []string
+	tlsConfig    *tls.Config
+	dispatcher   *dispatcher
+	authZPlugins []authorization.Plugin
+}
+
+// versionMatcher defines a variable matcher to be parsed by the router
+// when a request is about to be served.
+const versionMatcher = "/v{version:[0-9.]+}"
```
Comment (MHBauer) [inline api/server.go]: Good idea to make this a constant, but I don't see it used. Looks like it should be used in `api/primary.go`.

File: api/primary.go [hunk 3]
```diff
@@ -115,18 +119,23 @@ func profilerSetup(mainRouter *mux.Router, path string) {
...
 
 // NewPrimary creates a new API router.
-func NewPrimary(cluster cluster.Cluster, tlsConfig *tls.Config, status StatusHandler, debug, enableCors bool) *mux.Router {
+func NewPrimary(cluster cluster.Cluster, tlsConfig *tls.Config, status StatusHandler, debug, enableCors bool, authorizationPlugins []string) *mux.Router {
 	// Register the API events handler in the cluster.
 	eventsHandler := newEventsHandler()
 	cluster.RegisterEventHandler(eventsHandler)
 
-	context := &context{
+	context := &routerContext{
```
Comment (MHBauer) [inline api/primary.go]: Why was this renamed? Makes a lot of clutter in the diff.

**Comment by doronp** [comment]
@MHBauer There was a huge rebase maily  #1880  
Had lots of the same changes which were pre requisites for the AuthZ plugin usage.. 

That made separating Godeps and AuthZ plugin actual  usage less relevant.

About the next step of actually importing and using AuthZ plugin from the Engine:
I did try to test the authZ usage itself and I get conflict version of the discovery version used in it and in swarm:

cannot save github.com/docker/docker/pkg/authorization at revision 6eb38359954269103f2a3094283c4cb0a1cb63f9: already have github.com/docker/docker/pkg/discovery at revision b65fd8e879545e8c9b859ea9b6b825ac50c79e46.

Does that mean that docker authZ and swarm should decide on one revision? Its a bit of a blocker...

About the failing tests - I am not sure this is related to my changes.

File: api/primary.go [hunk 4]
```diff
@@ -7,6 +7,8 @@ import (
 	"net/http/pprof"
 
 	log "github.com/Sirupsen/logrus"
+	//	"github.com/docker/docker/pkg/authorization"
```
Comment (ezrasilvera) [inline api/primary.go]: Why don't you remove it for now instead of commenting out ?

**Comment by doronp** [comment]
@ezrasilvera Good catch. The comment out is due to Godeps indeed. Without that code there is no commit really, with it the error you get is misleading since I could not update the Godeps yet. I will try to determine if I can persuade one side to use the others version to resolve that.

**Comment by doronp** [comment]
So I changed the revision in the Swarm side it compiles now - Fails are not related to the change AFAIK.

**Comment by doronp** [comment]
@MHBauer Ping

**Comment by MHBauer** [comment]
Be forewarned, docker/docker#21897 is about to move the auth stuff around.

**Comment by duglin** [comment]
@vieux @abronan what do you think about the design of this?

**Comment by MHBauer** [comment]
I think the point of https://github.com/docker/swarm/pull/1366#issuecomment-164195664 was if we could just init the Middleware directly without copying the init into swarm. `newAuthorizationMiddleware` is already out of date due to changes in docker-engine. It appears to be just a copy of the initial version. If that was not the intent of the comment, it is still what I myself think.

**Comment by MHBauer** [comment]
We'd have to reach directly into `docker/docker/api/server/middleware` to do that though. (Un)Fortunately that is moved by docker/docker#21897  into a `pkg` so it will be easy to grab in the future. 

I can see that was considered, because it is commented out in `primary.go`. Some general cleanup to remove comments would be nice.

**Comment by doronp** [comment]
@MHBauer The reason newAuthorizationMiddleware is copied is that it is a bit different, Swarm handlers call chain does not support returning error whereas the dockers does.

In order to avoid that (almost a) copy there needs to be done another change of behaviour in Swarm which is that the handler being used should be able to return error and be like the one in Docker.

When that is done the behaviour could be the same as in Dockers middle wear. And then we can avoid that private function there and use Dockers.

I was planning to do that on another PR since it is not a small change.

**Comment by doronp** [comment]
@dongluochen - @abronan advised me you can help with this. What are the next steps on my side? 10x.

**Comment by hsluoyz** [comment]
Hi, may I ask what status of the ACL integration for Swarm is right now?

I have developed a Docker auth plugin called [casbin-authz-plugin](https://github.com/casbin/casbin-authz-plugin), and I want it to work on Docker Swarm too. What is the suggested way to do this? Thanks!

**Comment by dongluochen** [comment]
@hsluoyz Is your plugin for Docker Swarm-mode? You can find the difference of classic Swarm and Swarm-mode [here](https://github.com/docker/swarm#swarm-disambiguation). You may open issue at  https://github.com/moby/moby for Swarm mode authentication.

**Comment by hsluoyz** [comment]
Hi @dongluochen ,

> Is your plugin for Docker Swarm-mode?

No, my plugin is still a local Docker engine plugin base on: https://github.com/docker/go-plugins-helpers/tree/master/authorization

> You can find the difference of classic Swarm and Swarm-mode here.

So I think I should develop the auth plugin based on the new Swarm-mode on the moby repo.

> You may open issue at https://github.com/moby/moby for Swarm mode authentication.

I didn't find any information about how to develop an authorization plugin for Docker Swarm mode, I will open an issue in moby to ask about it as you suggested, thanks!
